### PR TITLE
Refactor rainflow HCM

### PR DESCRIPTION
**Pull request by johannes-mueller**
First step of refactoring FKM nonlinear rainflow counting

### CODE REVIEW

File: src/pylife/stress/rainflow/fkm_nonlinear.py [hunk 1]
```diff
@@ -247,45 +261,51 @@ def process(self, samples, flush=False):
...
-        if isinstance(samples, pd.DataFrame):
+        if multi_index:
             load_steps = samples.index.get_level_values('load_step').unique().to_series()
             if len(loads_indices) > 0:
                 tindex = loads_indices - old_head_index
                 idx = load_steps.iloc[tindex]
                 vals = samples.loc[idx].reset_index()
-                vals['load_step'].iloc[:len(self._last_sample)] = 0
                 vals = vals.set_index(['load_step', 'node_id'])
+                vals.columns = ['l']
```
[the remark this answers is not in the document]
Comment (maierbn) [reply src/pylife/stress/rainflow/fkm_nonlinear.py:283]: it's renamed in 298, can this be done from the start?

File: src/pylife/stress/rainflow/fkm_nonlinear.py [hunk 2]
```diff
@@ -446,56 +464,52 @@ def _proceed_on_secondary_branch(self, previous_point, current_point):
             The initially given current point, but with  updated values of stress and strain.
 
         """
-        delta_L = current_point.load - previous_point.load   # as described in FKM nonlinear
+        delta_L = current_point.load.values - previous_point.load.values   # as described in FKM nonlinear
+        index = current_point.load.index
+        index_levels = [n for n in index.names if n != 'load_step']
```
Comment (maierbn) [inline src/pylife/stress/rainflow/fkm_nonlinear.py:469]: rename to `obsolete_index_levels` or similar

File: src/pylife/stress/rainflow/fkm_nonlinear.py [hunk 3]
```diff
@@ -514,7 +528,8 @@ def _perform_hcm_algorithm(self, *, samples, recording_lists, largest_point, pre
         self._hcm_message += f"turning points: {samples}\n"
 
         # iterate over loads from the given list of samples
-        for current_load in load_turning_points:
+        li = load_turning_points.index.to_frame()['load_step']
+        for _, current_load in load_turning_points.groupby((li!=li.shift()).cumsum(), sort=False):
```
Comment (maierbn) [inline src/pylife/stress/rainflow/fkm_nonlinear.py:532]: The `li!=li.shift()` and `load_turning_points.groupby((li!=li.shift()).cumsum()` could be separate statements with variable assignment to make it better readable

File: src/pylife/stress/rainflow/recorders.py [hunk 4]
```diff
@@ -289,32 +288,32 @@ def collective(self):
...
+                    "S_max": self._S_max.to_numpy(),
+                    "R": self.R,
+                    "epsilon_min": self._epsilon_min.to_numpy(),
+                    "epsilon_max": self._epsilon_max.to_numpy(),
+                    "S_a": self.S_a,
+                    "S_m": self.S_m,
+                    "epsilon_a": self.epsilon_a,
+                    "epsilon_m": self.epsilon_m,
+                    "epsilon_min_LF": self._epsilon_min_LF.values,
+                    "epsilon_max_LF": self._epsilon_max_LF.to_numpy(),
```
Comment (maierbn) [inline src/pylife/stress/rainflow/recorders.py:313]: why `.values` once and the other time `.to_numpy()`, check if correct

**Review by maierbn — APPROVED** [review]
Please look for the comments, then it's fine

File: tests/stress/rainflow/test_new_turns.py [hunk 5]
```diff
@@ -224,6 +224,7 @@ def test_rainflow_new_turns_preserve_start(samples, expected_index, expected_val
 
 
 # with data frames
+@pytest.mark.skip(reason="handled in process")
```
[the remark this answers is not in the document]
Comment (johannes-mueller) [reply tests/stress/rainflow/test_new_turns.py:227]: I want to keep the tests around until the refactoring is actually done. Then we'll see, if we rewrite them or test the functionality differently.

File: src/pylife/stress/rainflow/fkm_nonlinear.py [hunk 6]
```diff
@@ -224,18 +228,28 @@ def process(self, samples, flush=False):
...
-        _S_min = []
-        _S_max = []
-        _epsilon_min = []
-        _epsilon_max = []
-        _epsilon_min_LF = []      # minimum strain of the load history up to (including) the current hysteresis (LF=Lastfolge), mentioned on p.127 of FKM nonlinear
-        _epsilon_max_LF = []      # maximum strain of the load history up to (including) the current hysteresis (LF=Lastfolge), mentioned on p.127 of FKM nonlinear
+        assert not isinstance(samples, pd.DataFrame)
+        multi_index = isinstance(samples, pd.Series) and len(samples.index.names) > 1
+
+        empty_index = [] if not isinstance(samples, pd.Series) else pd.DataFrame(columns=samples.index.names, dtype=np.int64).set_index(samples.index.names, drop=True).index
```
[the remark this answers is not in the document]
Comment (johannes-mueller) [reply src/pylife/stress/rainflow/fkm_nonlinear.py:234]: Good point. I think we will do a black reformatting once we release 2.1.1 and then enforce black formatting in the CI.

File: src/pylife/stress/rainflow/fkm_nonlinear.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (johannes-mueller) [reply src/pylife/stress/rainflow/fkm_nonlinear.py:283]: That's a relic from the time it all was in a `DataFrame`. Two lines later there is `load_turning_points = vals.iloc[:, 0]` so it actually does not matter what the name of the column is. I will delete the line and then proceed with the refactoring.

**Pull request merged**